src/menhir_parser.mly

```diff
@@ -59,31 +60,30 @@ let translate_old_rule : old_p_rule -> p_rule = fun r ->
             begin
               try
                 let n = Hashtbl.find arity x in
-                if nb_args < n then Hashtbl.replace arity x nb_args
+                if nb_args > n then Hashtbl.replace arity x nb_args
               with Not_found -> Hashtbl.add arity x nb_args
             end
       | P_Wild          -> ()
-      | P_Type          -> ()
+      | P_Type          -> fatal h.pos "Type in legacy pattern."
       | P_Prod(_,_)     -> fatal h.pos "Product in legacy pattern."
-      | P_Meta(_,_)     -> fatal h.pos "Metaviable in legacy pattern."
       | P_Abst(xs,t)    ->
           begin
             match xs with
             | [(_       ,Some(a),_)] ->
                 fatal a.pos "Annotation in legacy pattern."
             | [([Some x],None   ,_)] ->
                 compute_arities (x.elt::env) t
-            | _                      ->
-                fatal h.pos "Invalid legacy pattern lambda."
+            | _                      -> assert false
           end
-      | P_Patt(_,_)     -> fatal h.pos "Pattern in legacy rule."
       | P_Impl(_,_)     -> fatal h.pos "Implication in legacy pattern."
       | P_LLet(_,_,_,_) -> fatal h.pos "Let expression in legacy rule."
-      | P_NLit(_)       -> fatal h.pos "Nat literal in legacy rule."
-      | P_UnaO(_,_)     -> fatal h.pos "Unary operator in legacy rule."
-      | P_BinO(_,_,_)   -> fatal h.pos "Binary operator in legacy rule."
-      | P_Wrap(_)       -> fatal h.pos "Wrapping constructor in legacy rule."
-      | P_Expl(_)       -> fatal h.pos "Explicit argument in legacy rule."
+      | P_Meta(_,_)
+      | P_Patt(_,_)
+      | P_NLit(_)
+      | P_UnaO(_,_)
+      | P_BinO(_,_,_)
+      | P_Wrap(_)
+      | P_Expl(_)       -> assert false
     end;
```

Comment:
The changes done here are not a great idea: why remove all these decent error messages? Yes they may never happen, but a nice error message is better than an assert failure. Especially in the middle of this disgusting code that is very error-prone. Plus, this has nothing to do with the current PR.

Reply:
Because if this assertion fails then there is a bug. It's not a user error.

Reply:
Well, using the legacy syntax is a user error as far as I'm concerned... :)

Frankly, it is not a big deal. I think the previous code was better, but if you want to keep it that way then at least add an `assert false` for every matched pattern. You want to known which pattern triggered the error in the case of a bug.

Reply:
Fixed in cd1a1fd